File: Sources/WikiCore/WikiRequest.swift
//
//  WikiRequest.swift
//  WikiCore
//
//  Created by phoenix on 2025/12/17.
//

import Foundation
import Alamofire

// MARK: - Search Response Models

public struct WikiSearchResponse: Codable, Sendable {
    public let results: [WikiSearchResult]
    public let start: Int
    public let limit: Int
    public let totalSize: Int
    public let cqlQuery: String?
    public let searchDuration: Int?
    
    enum CodingKeys: String, CodingKey {
        case results, start, limit, totalSize, cqlQuery, searchDuration
    }
}

public struct WikiSearchResult: Codable, Sendable {
    public let content: WikiContent?
    public let title: String?
    public let excerpt: String?
    public let url: String?
    public let resultGlobalContainer: WikiContainer?
    public let entityType: String?
    public let iconCssClass: String?
    public let lastModified: String?
    public let friendlyLastModified: String?
}

public struct WikiContent: Codable, Sendable {
    public let id: String?
    public let type: String?
    public let status: String?
    public let title: String?
    public let _expandable: WikiExpandable?
    
    enum CodingKeys: String, CodingKey {
        case id, type, status, title, _expandable
    }
}

public struct WikiExpandable: Codable, Sendable {
    public let space: String?
}

public struct WikiContainer: Codable, Sendable {
    public let title: String?
    public let displayUrl: String?
}

// MARK: - Wiki API Client

public final class WikiAPIClient: @unchecked Sendable {
    
    public static let shared = WikiAPIClient()
    
    private let baseURL = "https://wiki.p1.cn"
    
    /// Cookie 管理器
    private let cookieManager = CookieManager.shared
    
    /// 日志记录器
    private var logger: WikiLoggerProtocol {
        WikiLoggerConfig.shared.logger
    }
    
    /// 获取请求头（包含动态获取的 Cookie）
    private var defaultHeaders: HTTPHeaders {
        var headers: HTTPHeaders = [:]
        
        if let cookie = cookieManager.getCookie() {
            headers["Cookie"] = cookie
        } else {
            logger.warning("无法获取 Cookie，API 请求可能会失败")
        }
        
        return headers
    }
    
    private init() {
        // 初始化时检查 Cookie 状态
        if cookieManager.getCookie() == nil || cookieManager.getCookie()?.isEmpty == true {
            logger.warning("未配置 Cookie，请设置环境变量 WIKI_COOKIE")
        }
    }
    
    // MARK: - Search API
    
    /// 搜索 Wiki 内容
    /// - Parameters:
    ///   - query: 搜索关键词
    ///   - start: 起始位置，默认 0
    ///   - limit: 返回数量限制，默认 20
    /// - Returns: WikiSearchResponse
    public func search(
        query: String,
        start: Int = 0,
        limit: Int = 20
    ) async throws -> WikiSearchResponse {
        let url = "\(baseURL)/rest/api/search"
        
        let parameters: Parameters = [
            "cql": "siteSearch ~ \"\(query)\" AND type in (\"space\",\"user\",\"page\",\"blogpost\",\"attachment\")",
            "start": "\(start)",
            "limit": "\(limit)",
            "excerpt": "highlight",
            "expand": "space.icon",
            "includeArchivedSpaces": "false",
            "src": "next.ui.search"
        ]
        
        return try await withCheckedThrowingContinuation { continuation in
            AF.request(url, method: .get, parameters: parameters, headers: defaultHeaders)
                .validate(statusCode: 200..<300)
                .responseDecodable(of: WikiSearchResponse.self) { response in
                    switch response.result {
                    case .success(let searchResponse):
                        continuation.resume(returning: searchResponse)
                    case .failure(let error):
                        continuation.resume(throwing: error)
                    }
                }
        }
    }
    
    // MARK: - View Page API
    
    /// 获取页面 HTML 内容
    /// - Parameter pageId: 页面 ID
    /// - Returns: 页面 HTML 字符串
    public func viewPage(pageId: String) async throws -> String {
        let url = "\(baseURL)/pages/viewpage.action"
        
        let parameters: Parameters = [
            "pageId": pageId
        ]
        
        return try await withCheckedThrowingContinuation { continuation in
            AF.request(url, method: .get, parameters: parameters, headers: defaultHeaders)
                .validate(statusCode: 200..<300)
                .responseString { response in
                    switch response.result {
                    case .success(let html):
                        continuation.resume(returning: html)
                    case .failure(let error):
                        continuation.resume(throwing: error)
                    }
                }
        }
    }
    
    /// 获取页面 HTML 内容
    /// - Parameter url: 页面地址
    /// - Returns: 页面 HTML 字符串
    public func viewPage(url: String) async throws -> String {
        
        return try await withCheckedThrowingContinuation { continuation in
            AF.request(url, method: .get, parameters: nil, headers: defaultHeaders)
                .validate(statusCode: 200..<300)
                .responseString { response in
                    switch response.result {
                    case .success(let html):
                        continuation.resume(returning: html)
                    case .failure(let error):
                        continuation.resume(throwing: error)
                    }
                }
        }
    }
    
    /// 通过完整 URL 下载图片
    /// - Parameter url: 完整的图片 URL
    /// - Returns: 图片 Data
    public func downloadImage(from url: String) async throws -> Data {
        return try await withCheckedThrowingContinuation { continuation in
            AF.request(url, method: .get, headers: defaultHeaders)
                .validate(statusCode: 200..<300)
                .responseData { response in
                    switch response.result {
                    case .success(let data):
                        continuation.resume(returning: data)
                    case .failure(let error):
                        continuation.resume(throwing: error)
                    }
                }
        }
    }
}

// MARK: - Error Types

public enum WikiAPIError: Error, LocalizedError {
    case invalidURL
    case noData
    case decodingFailed(Error)
    case networkError(Error)
    
    public var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .noData:
            return "No data received"
        case .decodingFailed(let error):
            return "Decoding failed: \(error.localizedDescription)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        }
    }
}

File: skills/wiki-tool/wiki_request.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Wiki API 请求模块

提供 Confluence Wiki API 的访问能力：
- 搜索 Wiki 内容
- 获取页面 HTML
- 下载图片
"""

import os
import sys
import warnings
from dataclasses import dataclass, field
from typing import Optional, List, Dict, Any

import requests
from requests.packages.urllib3.exceptions import InsecureRequestWarning

# 禁用 SSL 警告（内网服务器可能使用自签名证书）
warnings.simplefilter('ignore', InsecureRequestWarning)


# MARK: - Data Models

@dataclass
class WikiExpandable:
    """Wiki 可扩展字段"""
    space: Optional[str] = None


@dataclass
class WikiContent:
    """Wiki 内容信息"""
    id: Optional[str] = None
    type: Optional[str] = None
    status: Optional[str] = None
    title: Optional[str] = None
    expandable: Optional[WikiExpandable] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'WikiContent':
        expandable = None
        if '_expandable' in data:
            expandable = WikiExpandable(space=data['_expandable'].get('space'))
        return cls(
            id=data.get('id'),
            type=data.get('type'),
            status=data.get('status'),
            title=data.get('title'),
            expandable=expandable
        )


@dataclass
class WikiContainer:
    """Wiki 容器信息"""
    title: Optional[str] = None
    display_url: Optional[str] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'WikiContainer':
        return cls(
            title=data.get('title'),
            display_url=data.get('displayUrl')
        )


@dataclass
class WikiSearchResult:
    """Wiki 搜索结果项"""
    content: Optional[WikiContent] = None
    title: Optional[str] = None
    excerpt: Optional[str] = None
    url: Optional[str] = None
    result_global_container: Optional[WikiContainer] = None
    entity_type: Optional[str] = None
    icon_css_class: Optional[str] = None
    last_modified: Optional[str] = None
    friendly_last_modified: Optional[str] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'WikiSearchResult':
        content = None
        if 'content' in data and data['content']:
            content = WikiContent.from_dict(data['content'])
        
        container = None
        if 'resultGlobalContainer' in data and data['resultGlobalContainer']:
            container = WikiContainer.from_dict(data['resultGlobalContainer'])
        
        return cls(
            content=content,
            title=data.get('title'),
            excerpt=data.get('excerpt'),
            url=data.get('url'),
            result_global_container=container,
            entity_type=data.get('entityType'),
            icon_css_class=data.get('iconCssClass'),
            last_modified=data.get('lastModified'),
            friendly_last_modified=data.get('friendlyLastModified')
        )


@dataclass
class WikiSearchResponse:
    """Wiki 搜索响应"""
    results: List[WikiSearchResult] = field(default_factory=list)
    start: int = 0
    limit: int = 20
    total_size: int = 0
    cql_query: Optional[str] = None
    search_duration: Optional[int] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'WikiSearchResponse':
        results = [WikiSearchResult.from_dict(r) for r in data.get('results', [])]
        return cls(
            results=results,
            start=data.get('start', 0),
            limit=data.get('limit', 20),
            total_size=data.get('totalSize', 0),
            cql_query=data.get('cqlQuery'),
            search_duration=data.get('searchDuration')
        )


# MARK: - API Client

class WikiAPIClient:
    """Wiki API 客户端"""
    
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if self._initialized:
            return
        
        self.base_url = "https://wiki.p1.cn"
        self._cookie = os.environ.get('WIKI_COOKIE')
        self._session = requests.Session()
        self._session.verify = False  # 禁用 SSL 证书验证
        
        if not self._cookie:
            print("警告: 未设置 WIKI_COOKIE 环境变量，API 请求可能会失败", file=sys.stderr)
        
        self._initialized = True
    
    @property
    def _headers(self) -> Dict[str, str]:
        """获取请求头"""
        headers = {}
        if self._cookie:
            headers['Cookie'] = self._cookie
        return headers
    
    # MARK: - Search API
    
    def search(self, query: str, start: int = 0, limit: int = 20) -> WikiSearchResponse:
        """
        搜索 Wiki 内容
        
        Args:
            query: 搜索关键词
            start: 起始位置，默认 0
            limit: 返回数量限制，默认 20
            
        Returns:
            WikiSearchResponse
        """
        url = f"{self.base_url}/rest/api/search"
        
        params = {
            'cql': f'siteSearch ~ "{query}" AND type in ("space","user","page","blogpost","attachment")',
            'start': str(start),
            'limit': str(limit),
            'excerpt': 'highlight',
            'expand': 'space.icon',
            'includeArchivedSpaces': 'false',
            'src': 'next.ui.search'
        }
        
        response = self._session.get(url, params=params, headers=self._headers)
        response.raise_for_status()
        
        return WikiSearchResponse.from_dict(response.json())
    
    # MARK: - View Page API
    
    def view_page_by_id(self, page_id: str) -> str:
        """
        通过页面 ID 获取 HTML 内容
        
        Args:
            page_id: 页面 ID
            
        Returns:
            页面 HTML 字符串
        """
        url = f"{self.base_url}/pages/viewpage.action"
        params = {'pageId': page_id}
        
        response = self._session.get(url, params=params, headers=self._headers)
        response.raise_for_status()
        
        return response.text
    
    def view_page_by_url(self, page_url: str) -> str:
        """
        通过 URL 获取页面 HTML 内容
        
        Args:
            page_url: 页面完整 URL
            
        Returns:
            页面 HTML 字符串
        """
        response = self._session.get(page_url, headers=self._headers)
        response.raise_for_status()
        
        return response.text
    
    # MARK: - Image Download
    
    def download_image(self, url: str) -> bytes:
        """
        下载图片
        
        Args:
            url: 图片完整 URL
            
        Returns:
            图片二进制数据
        """
        response = self._session.get(url, headers=self._headers)
        response.raise_for_status()
        
        return response.content


# MARK: - Error Types

class WikiAPIError(Exception):
    """Wiki API 错误基类"""
    pass


class InvalidURLError(WikiAPIError):
    """无效 URL 错误"""
    pass


class NoDataError(WikiAPIError):
    """无数据错误"""
    pass


class NetworkError(WikiAPIError):
    """网络错误"""
    def __init__(self, original_error: Exception):
        self.original_error = original_error
        super().__init__(f"网络错误: {original_error}")


# 单例实例
wiki_client = WikiAPIClient()

